lib/SPIRV/SPIRVBuiltinHelper.h

```diff
@@ -0,0 +1,279 @@
+  llvm::Value *getMutated() { return doConversion(); }
+
+  /// Return the number of arguments currently specified for the new call.
+  unsigned arg_size() const { return Args.size(); }
+
+  /// Get the corresponding argument for the new call.
+  llvm::Value *getArg(unsigned Index) const { return Args[Index]; }
+
+  /// Return the pointer element type of the corresponding index, or nullptr if
+  /// it is not a pointer.
+  llvm::Type *getPointerElementType(unsigned Index) const {
+    llvm::TypedPointerType *ElTy = PointerTypes[Index];
+    return ElTy ? ElTy->getElementType() : nullptr;
+  }
+
+  /// A pair representing both the LLVM value of an argument and its
+  /// corresponding pointer element type. This type can be constructed from
+  /// implicit conversion from an LLVM value object (but only if it is not of
+  /// pointer type), or by the appropriate std::pair type.
+  struct ValueTypePair : public std::pair<llvm::Value *, llvm::Type *> {
+    ValueTypePair(llvm::Value *V) : pair(V, nullptr) {
+      assert(!V->getType()->isPointerTy() &&
+             "Must specify a pointer element type if value is a pointer.");
+    }
+    ValueTypePair(std::pair<llvm::Value *, llvm::Type *> P) : pair(P) {}
+    ValueTypePair(llvm::Value *V, llvm::Type *T) : pair(V, T) {}
+    ValueTypePair() = delete;
+    using pair::pair;
+  };
+
+  /// Use the following arguments as the arguments of the new call, replacing
+  /// any previous arguments. This version may not be used if any argument is of
+  /// pointer type.
+  BuiltinCallMutator &setArgs(llvm::ArrayRef<llvm::Value *> Args);
+
+  /// This will replace the return type of the call with a different return
+  /// type. The second argument is a function that will be called with an
+  /// IRBuilder parameter and the newly generated function, and will return the
+  /// value to replace all uses of the original call instruction with. Example
+  /// usage:
+  ///
+  ///     BuiltinCallMutator Mutator = /* ... */;
+  ///     Mutator.changeReturnType(Int16Ty, [](IRBuilder<> &IRB, CallInst *CI) {
+  ///       return IRB.CreateZExt(CI, Int16Ty);
+  ///     });
+  BuiltinCallMutator &changeReturnType(llvm::Type *ReturnTy,
+                                       MutateRetFuncTy MutateFunc);
+
+  /// Insert an argument before the given index.
+  BuiltinCallMutator &insertArg(unsigned Index, ValueTypePair Arg);
+
+  /// Add an argument to the end of the argument list.
+  BuiltinCallMutator &appendArg(ValueTypePair Arg) {
+    return insertArg(Args.size(), Arg);
+  }
+
+  /// Replace the argument at the given index with a new value.
+  BuiltinCallMutator &replaceArg(unsigned Index, ValueTypePair Arg);
+
+  /// Remove the argument at the given index.
+  BuiltinCallMutator &removeArg(unsigned Index);
+
+  /// Remove all arguments in a range.
+  BuiltinCallMutator &removeArgs(unsigned Start, unsigned Len) {
+    for (unsigned I = 0; I < Len; I++)
+      removeArg(Start);
```

Comment:
Or actually from Start to Start + Len. Or am I missing something?

Reply:
I'm not entirely sure what you mean here. The contract of the method is to remove arguments [`Start`, `Start+Len`), so `for (unsigned I = Start; I < Len; I++) removeArg(I);` would do a very different, and wrong, thing.

Reply:
`removeArg(I);` in lieu of `removeArg(Start);` would cause the wrong arguments to be removed, since after calling the function once, arguments [`Start+1`..] get renumbered to [`Start`..]. As a result, using `removeArg(I);` instead would cause, say, arguments 2, 4, 6, 8, etc. to be removed instead of 2, 3, 4, 5. This can be fixed by doing reverse iteration, but that's also somewhat trickier to code correctly, and I opted for the simpler implementation.

Reply:
Ah, got it. Thanks!